### PR TITLE
[ssl] Use ring buffer in ssl transport layer

### PR DESCRIPTION
**Pull request by euroelessar**
BIO_pair is implemented on top ring buffer which is more efficient compared to current BIO_mem approach.

See #14058

### CODE REVIEW

**Comment by thelinuxfoundation** [comment]
Thank you for your pull request. Before we can look at your contribution, we need to ensure all contributors are covered by a Contributor License Agreement.

After the following items are addressed, please respond with a new comment here, and the automated system will re-verify.

* User @euroelessar isn't covered by a CLA. They will need to complete the form at https://identity.linuxfoundation.org/projects/cncf

_Regards,
CLA GitHub bot_

**Comment by euroelessar** [comment]
Performed cla signup process (hint for bot)

**Comment by euroelessar** [comment]
@jboeuf @justinburke can you have a look please?

**Comment by jboeuf** [comment]
I like the idea. I will let @jiangtaoli2016 do the review.

File: src/core/tsi/ssl_transport_security.cc [hunk 1]
```diff
@@ -1107,16 +1107,12 @@ static tsi_result create_tsi_ssl_handshaker(SSL_CTX* ctx, int is_client,
...
 
-  into_ssl = BIO_new(BIO_s_mem());
-  from_ssl = BIO_new(BIO_s_mem());
-  if (into_ssl == nullptr || from_ssl == nullptr) {
-    gpr_log(GPR_ERROR, "BIO_new failed.");
+  if (!BIO_new_bio_pair(&network_io, 0, &ssl_io, 0)) {
+    gpr_log(GPR_ERROR, "BIO_new_bio_pair failed.");
     SSL_free(ssl);
-    if (into_ssl != nullptr) BIO_free(into_ssl);
-    if (from_ssl != nullptr) BIO_free(into_ssl);
```
Comment (SaveTheRbtz) [inline src/core/tsi/ssl_transport_security.cc]: image

File: src/core/tsi/ssl_transport_security.cc [hunk 2]
```diff
@@ -96,17 +96,15 @@ struct tsi_ssl_server_handshaker_factory {
 typedef struct {
   tsi_handshaker base;
   SSL* ssl;
-  BIO* into_ssl;
-  BIO* from_ssl;
+  BIO *network_io;
```
Comment (jiangtaoli2016) [inline src/core/tsi/ssl_transport_security.cc]: Please use "BIO* network_io;" to be consistent with the existing coding style.

File: src/core/tsi/ssl_transport_security.cc [hunk 3]
```diff
@@ -1062,8 +1061,8 @@ static tsi_result ssl_handshaker_create_frame_protector(
    * cannot call anything else but destroy on the handshaker after this call. */
```
Comment (jiangtaoli2016) [inline src/core/tsi/ssl_transport_security.cc]: Please also update comment that transfer ownership of ssl and network_io to frame protector.

**Comment by euroelessar** [comment]
@jiangtaoli2016 thanks, I've addressed comments

**Comment by jiangtaoli2016** [comment]
@euroelessar Thanks for revision. Looks good to me. Let's wait for test result.

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                              FILE SIZE
 ++++++++++++++ GROWING                                ++++++++++++++
  +0.0%     +64 [None]                                 +3.38Ki  +0.1%
  +0.1%     +16 src/core/tsi/ssl_transport_security.cc     +16  +0.1%
      +4.5%     +14 [Unmapped]                                 +14  +4.5%
       +34%     +14 ssl_protector_destroy                      +14   +34%
       +29%      +9 ssl_handshaker_destroy                      +9   +29%

  +0.0%     +80 TOTAL                                  +3.39Ki  +0.1%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Pull request merged**